src/sqlfluff/rules/L003.py

```diff
@@ -772,8 +772,9 @@ def _handle_template_blocks(
         # both have lines where anchor_indent_balance drops 2 levels from one line
         # to the next, making it a bit unclear how to indent that line.
         template_line = _find_matching_start_line(previous_lines)
-        # This cant occur in valid code
-        assert template_line, "TypeGuard"
+        # In rare circumstances there may be disbalanced pairs
+        if not template_line:
+            return LintResult(memory=memory)
```

Comment:
It looks like this will avoid the exception. Should we consider looking at Jinja brackets in "source space" rather than "templated space"? The stripped whitespace is still present there.

I don't have a strong feeling about this, just thought I'd mention it. Some rules do this for various reasons (including L003 -- it distinguishes between source and templated whitespace for indentation, ignoring templated whitespace because it's not "real" from a user perspective).

Reply:
I checked the other rules and it seems ok. I believe I implemented this check during the refactor.
the consideration with the assert was that it shouldn't be possible for us to have got this far if the jinja tags are misbalanced, However I think that was never true. A pretty special thing is happening here - The line is actually on its own, allowing it to get past the first check, however the -%} pulls it on to the same line which makes it disbalanced

Reply:
That's fine. We've gradually added this kind of logic to some rules. It's often a judgement call whether it's worth the effort. We can always revisit it later, depending on what our users report, how busy we are, etc.